[JSC] add support for `Iterator.prototype.chunks` and `Iterator.prototype.windows`

dcrousso opened the pull request:
#### 6d98a9cf99d4e000702b64ca31e9f5eefd413e0f
<pre>
[JSC] add support for `Iterator.prototype.chunks` and `Iterator.prototype.windows`
<a href="https://bugs.webkit.org/show_bug.cgi?id=281112">https://bugs.webkit.org/show_bug.cgi?id=281112</a>

Reviewed by Alexey Shvayka.

These will allow for alternative consumptions of iterators (e.g. batched processing, something like C++ `std::adjacent_difference`, etc.) instead of having to manually create a generator wrapper.

Proposal: &lt;<a href="https://github.com/tc39/proposal-iterator-chunking">https://github.com/tc39/proposal-iterator-chunking</a>&gt;
Spec: &lt;<a href="https://tc39.es/proposal-iterator-chunking/">https://tc39.es/proposal-iterator-chunking/</a>&gt;

* Source/JavaScriptCore/builtins/JSIteratorPrototype.js:
(chunks): Added.
(windows): Added.
* Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
(JSC::JSIteratorPrototype::finishCreation):
* Source/JavaScriptCore/runtime/OptionsList.h:

* JSTests/stress/iterator-prototype-chunks.js: Added.
* JSTests/stress/iterator-prototype-windows.js: Added.

Canonical link: <a href="https://commits.webkit.org/285240@main">https://commits.webkit.org/285240@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c30f399c48748b13d60a9f9695e86d8ac0b6becb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70991 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50403 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23764 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75095 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22196 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58202 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22014 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56149 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14623 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74057 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45791 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61195 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36595 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42442 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18629 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20537 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/64115 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64383 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18989 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76811 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/70240 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15223 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18186 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63878 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15267 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61254 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63839 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15905 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11933 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5581 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/92021 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46204 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20058 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47276 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48559 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47018 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/JavaScriptCore/builtins/JSIteratorPrototype.js:
```diff
@@ -310,3 +310,93 @@ function reduce(reducer /*, initialValue */)
+                var oldBuffer = buffer;
+                buffer = @newArrayWithSize(oldBuffer.length - 1);
+                for (var i = 0; i < buffer.length; ++i)
+                    buffer[i] = oldBuffer[i + 1];
```
note that this isn't currently accurate with what's in the (proposed) spec as it seems potentially not ideal <https://github.com/tc39/proposal-iterator-chunking/issues/11> <[https://webkit.slack.com/archives/CTV4FGWF4/p1728442953003169](https://webkit.slack.com/archives/CTV4FGWF4/p1728442953003169)>

shvaikalesh commented on Source/JavaScriptCore/builtins/JSIteratorPrototype.js:
```diff
@@ -310,3 +310,93 @@ function reduce(reducer /*, initialValue */)
+        @throwRangeError("Iterator.prototype.chunks requires that argument not be NaN.");
+
+    var intChunkSize = @toIntegerOrInfinity(numChunkSize);
+    if (intChunkSize < 1 || intChunkSize > 0xffffffff)
```
Would be nice to either utilize `@MAX_ARRAY_INDEX` somehow or add a new constant.

shvaikalesh commented on Source/JavaScriptCore/builtins/JSIteratorPrototype.js:
```diff
@@ -310,3 +310,93 @@ function reduce(reducer /*, initialValue */)
+            if (result.done) {
+                if (buffer.length)
+                    yield buffer;
+                break;
```
```suggestion
                return;
```

shvaikalesh commented on Source/JavaScriptCore/builtins/JSIteratorPrototype.js:
```diff
@@ -310,3 +310,93 @@ function reduce(reducer /*, initialValue */)
+        for (;;) {
+            var result = @iteratorGenericNext(iteratedNextMethod, iterated);
+            if (result.done)
+                break;
```
Let's do `return` here as well to match the spec and other `Iterator.prototype` methods.

shvaikalesh reviewed (changes requested):
nits

sosukesuzuki commented on Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
```diff
@@ -83,6 +83,10 @@ void JSIteratorPrototype::finishCreation(VM& vm, JSGlobalObject* globalObject)
         JSC_BUILTIN_FUNCTION_WITHOUT_TRANSITION("drop"_s, jsIteratorPrototypeDropCodeGenerator, static_cast<unsigned>(PropertyAttribute::DontEnum));
         // https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.flatmap
         JSC_BUILTIN_FUNCTION_WITHOUT_TRANSITION(vm.propertyNames->builtinNames().flatMapPublicName(), jsIteratorPrototypeFlatMapCodeGenerator, static_cast<unsigned>(PropertyAttribute::DontEnum));
+        // https://tc39.es/proposal-iterator-chunking/#sec-iterator.prototype.chunks
+        JSC_BUILTIN_FUNCTION_WITHOUT_TRANSITION("chunks"_s, jsIteratorPrototypeChunksCodeGenerator, static_cast<unsigned>(PropertyAttribute::DontEnum));
+        // https://tc39.es/proposal-iterator-chunking/#sec-iterator.prototype.windows
+        JSC_BUILTIN_FUNCTION_WITHOUT_TRANSITION("windows"_s, jsIteratorPrototypeWindowsCodeGenerator, static_cast<unsigned>(PropertyAttribute::DontEnum));
```
I think we should add new feature flag like a `useIteratorChunking`. This is because `IteratorHelpers` and `IteratorChunking` are different proposals and should be enabled by default at different times.

(But I'm not a WebKit Reviewer so this is an unofficial review)

shvaikalesh reviewed (approved):
Looks great, thank you!

dcrousso committed:
[JSC] add support for `Iterator.prototype.chunks` and `Iterator.prototype.windows`
https://bugs.webkit.org/show_bug.cgi?id=281112

Reviewed by Alexey Shvayka.

These will allow for alternative consumptions of iterators (e.g. batched processing, something like C++ `std::adjacent_difference`, etc.) instead of having to manually create a generator wrapper.

Proposal: <https://github.com/tc39/proposal-iterator-chunking>
Spec: <https://tc39.es/proposal-iterator-chunking/>

* Source/JavaScriptCore/builtins/JSIteratorPrototype.js:
(chunks): Added.
(windows): Added.
* Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
(JSC::JSIteratorPrototype::finishCreation):
* Source/JavaScriptCore/runtime/OptionsList.h:

* JSTests/stress/iterator-prototype-chunks.js: Added.
* JSTests/stress/iterator-prototype-windows.js: Added.

Canonical link: https://commits.webkit.org/285240@main
